src/link-preview-expander.ts

```diff
@@ -1,16 +1,75 @@
 import * as sourcegraph from 'sourcegraph'
+import { concat, of } from 'rxjs'
+import { fromFetch } from 'rxjs/fetch'
+import { catchError, map, switchMap } from 'rxjs/operators'
 
-export function activate(ctx: sourcegraph.ExtensionContext): void {
-    ctx.subscriptions.add(
+export function activate(context: sourcegraph.ExtensionContext): void {
+    context.subscriptions.add(
         sourcegraph.languages.registerHoverProvider(['*'], {
-            provideHover: () => ({
-                contents: {
-                    value: 'Hello world from link-preview-expander! 🎉🎉🎉',
-                    kind: sourcegraph.MarkupKind.Markdown
+            provideHover: (document, position) => {
+                const range = document.getWordRangeAtPosition(position)
+                if (!range) {
+                    return null
                 }
-            }),
+                const maybeURL = getWord(document, range)
+                const isURL = checkIsURL(maybeURL)
+
+                if (!isURL) {
+                    return null
+                }
+
+                const createResult: (debugString: string) => sourcegraph.Badged<sourcegraph.Hover> = (
+                    debugString: string
+                ) => ({
+                    contents: {
+                        value: `<p>${debugString}</p>
+                        <a href="${maybeURL}" target="_blank" rel="noopener noreferrer">Navigate to link</a>
+                        <img width="64" src="https://miro.medium.com/max/816/1*mn6bOs7s6Qbao15PMNRyOA.png" />`,
+                        kind: sourcegraph.MarkupKind.Markdown,
+                    },
+                })
+
+                return concat(
+                    of(createResult('loading')),
```

Comment:
The extension host will display a loader automatically as long as the provider has not emitted a result, so you don't need to emit something like this :)

Reply:
I did this so that links can be clickable before fetching open graph metadata. Most links will not have previews, so I didn't want to slow those hovers down. We could display the image above the link to prevent layout shift from moving the link away from the users pointer.

Reply:
You could also register two separate hover providers, or make following the link the "go to definition" action